storage/index/types.go

```diff
@@ -183,42 +179,306 @@ type BlockTickResult struct {
+type writeBatchSortBy uint
+
+const (
+	writeBatchSortByUnmarkedAndBlockStart writeBatchSortBy = iota
+	writeBatchSortByEnqueued
+)
+
+// WriteBatchOptions is a set of options required for a write batch.
+type WriteBatchOptions struct {
+	InitialCapacity int
+	IndexBlockSize  time.Duration
+}
+
+// NewWriteBatch creates a new write batch.
+func NewWriteBatch(opts WriteBatchOptions) *WriteBatch {
+	return &WriteBatch{
+		opts:    opts,
+		entries: make([]WriteBatchEntry, 0, opts.InitialCapacity),
+		docs:    make([]doc.Document, 0, opts.InitialCapacity),
+	}
+}
+
+// Append appends an entry with accompanying document.
+func (b *WriteBatch) Append(
+	entry WriteBatchEntry,
+	doc doc.Document,
+) {
+	// Set private WriteBatchEntry fields
+	entry.enqueuedIdx = len(b.entries)
+	entry.result = WriteBatchEntryResult{}
+
+	// Append
+	b.entries = append(b.entries, entry)
+	b.docs = append(b.docs, doc)
+}
+
+// ForEachWriteBatchEntryFn allows a caller to perform an operation for each
+// batch entry.
+type ForEachWriteBatchEntryFn func(
+	idx int,
+	entry WriteBatchEntry,
+	doc doc.Document,
+	result WriteBatchEntryResult,
+)
+
+// ForEach allows a caller to perform an operation for each batch entry.
+func (b *WriteBatch) ForEach(fn ForEachWriteBatchEntryFn) {
+	for idx, entry := range b.entries {
+		fn(idx, entry, b.docs[idx], entry.Result())
+	}
+}
+
+// ForEachWriteBatchByBlockStartFn allows a caller to perform an operation with
+// reference to a restricted set of the write batch for each unique block
+// start.
+type ForEachWriteBatchByBlockStartFn func(
+	blockStart time.Time,
+	batch *WriteBatch,
+)
+
+// ForEachUnmarkedBatchByBlockStart allows a caller to perform an operation
+// with reference to a restricted set of the write batch for each unique block
+// start for entries that have not been marked completed yet.
+func (b *WriteBatch) ForEachUnmarkedBatchByBlockStart(
+	fn ForEachWriteBatchByBlockStartFn,
+) {
+	// Ensure sorted correctly first
+	b.SortByUnmarkedAndIndexBlockStart()
+
+	// What we do is a little funky but least alloc intensive, essentially we mutate
+	// this batch and then restore the pointers to the original docs after.
+	allEntries := b.entries
+	allDocs := b.docs
+	defer func() {
+		b.entries = allEntries
+		b.docs = allDocs
+	}()
 
-// ForEachBlockStart iterates over the provided WriteBatchEntryByBlockStart, and calls `fn` on each
-// group of elements with the same blockStart.
-func (w WriteBatchEntryByBlockStart) ForEachBlockStart(fn ForEachBlockStartFn) {
 	var (
-		startIdx  = 0
-		lastNanos xtime.UnixNano
+		blockSize      = b.opts.IndexBlockSize
+		startIdx       = 0
+		lastBlockStart xtime.UnixNano
 	)
-	for i := 0; i < len(w); i++ {
-		elem := w[i]
-		if elem.BlockStart != lastNanos {
-			lastNanos = elem.BlockStart
-			// We only want to call the the ForEachBlockStartFn once we have calculated the entire group,
+	for i := range allEntries {
+		if allEntries[i].OnIndexSeries == nil {
```

Comment:
making sure i understand this - you can early terminate here because the sort above guarantees if the first entry is done, then all are done, yea?

Reply:
Right yeah.